File: go/strings/reader.go
/*
 * Copyright (C) 2021 Massimiliano Ghilardi
 *
 *     This Source Code Form is subject to the terms of the Mozilla Public
 *     License, v. 2.0. If a copy of the MPL was not distributed with this
 *     file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 *
 * reader.go
 *
 *  Created on: Apr 07, 2021
 *      Author: Massimiliano Ghilardi
 */

package strings

import (
	"github.com/cosmos72/onejit/go/io"
)

// Reduced version of strings.Reader, implements only io.Reader
// by reading from a string.
// The zero value is a Reader of an empty string.
type Reader struct {
	str string
}

func (r *Reader) Read(p []byte) (n int, err error) {
	str := r.str
	n = len(str)
	if n == 0 {
		return 0, io.EOF
	}
	copy(p, str)
	if n > len(p) {
		n = len(p)
	}
	r.str = str[n:]
	return n, nil
}

func (r *Reader) Reset(str string) {
	r.str = str
}

// return a new Reader reading from specified string str.
func NewReader(str string) *Reader {
	return &Reader{str}
}

File: onestl/io/reader.cpp
/*
 * onestl - tiny STL C++ library
 *
 * Copyright (C) 2018-2021 Massimiliano Ghilardi
 *
 *     This Source Code Form is subject to the terms of the Mozilla Public
 *     License, v. 2.0. If a copy of the MPL was not distributed with this
 *     file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * reader.cpp
 *
 *  Created on Jul 09, 2022
 *      Author Massimiliano Ghilardi
 */

#include <onestl/io/reader.hpp>

namespace onestl {
namespace io {

////////////////////////////////////////////////////////////////////////////////
int Reader::read(char *chars, size_t n) const {
  int err = 0;
  if (func_) {
    err = func_(handle_, offset_, chars, n);
    if (err > 0) {
      offset_ += (uint64_t)err;
    }
  }
  return err;
}

} // namespace io
} // namespace onestl
